Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -15,709 +15,573 @@
-        let version = HTTPVersion(major: Int(parser.pointee.http_major), minor: Int(parser.pointee.http_minor))
-        let request = HTTPRequestHead(version: version, method: method, rawURI: state.currentURI!, headers: HTTPHeaders(buffer: cumulationBuffer!, headers: state.currentHeaders, keepAliveState: parser.keepAliveState))
-        self.state.currentHeaders.removeAll(keepingCapacity: true)
-        return request
+    private func finish(_ callout: (inout HTTPDecoderDelegate, UnsafeRawBufferPointer) -> Void) {
+        var currentFieldByteLength = 0
+        swap(&currentFieldByteLength, &self.currentFieldByteLength)
+        let start = self.rawBytesView.startIndex + self.firstNonDiscardableOffset!
+        let end = self.rawBytesView.startIndex + self.firstNonDiscardableOffset! + currentFieldByteLength
+        self.firstNonDiscardableOffset = nil
+        callout(&self.delegate, .init(rebasing: self.rawBytesView[start ..< end]))
     }
 
-    private func newResponseHead(_ parser: UnsafeMutablePointer<http_parser>!) -> HTTPResponseHead {
-        let status = HTTPResponseStatus(statusCode: Int(parser.pointee.status_code), reasonPhrase: state.currentStatus!)
-        let version = HTTPVersion(major: Int(parser.pointee.http_major), minor: Int(parser.pointee.http_minor))
-        let response = HTTPResponseHead(version: version, status: status, headers: HTTPHeaders(buffer: cumulationBuffer!, headers: state.currentHeaders, keepAliveState: parser.keepAliveState))
-        self.state.currentHeaders.removeAll(keepingCapacity: true)
-        return response
+    private func didReceiveBodyData(_ bytes: UnsafeRawBufferPointer) {
+        self.delegate.didReceiveBody(bytes)
     }
 
-    private func bytesToForwardOnRemoval(context: ChannelHandlerContext) -> ByteBuffer? {
-        guard self.leftOverBytesStrategy == .forwardBytes && self.parser.upgrade == 1 && context.channel.isActive else {
-            return nil
-        }
-        // We take a slice of the cumulationBuffer so the next handler in the pipeline will just see the readable portion of the buffer.
-        // While this is not strictly needed it may make it easier to consume.
-        if let buffer = self.cumulationBuffer?.slice(), buffer.readableBytes > 0 {
-            return buffer
+    private func didReceiveHeaderFieldData(_ bytes: UnsafeRawBufferPointer) {
+        switch self.decodingState {
+        case .headerName, .trailerName:
+            ()
+        case .headerValue:
+            self.finish { decoder, bytes in
+                decoder.didReceiveHeaderValue(bytes)
+            }
+            self.start(bytes: bytes, newState: .headerName)
+        case .trailerValue:
+            self.finish { decoder, bytes in
+                decoder.didReceiveTrailerValue(bytes)
+            }
+            self.start(bytes: bytes, newState: .trailerName)
+        case .url:
+            self.finish { decoder, bytes in
```

Comment:
This is a very minor nit that you can ignore, but all of these blocks call the first argument `decoder`, while on `self` it's called `delegate`.

Reply:
good point, will reanem